src/charm.py

```diff
@@ -141,22 +143,46 @@ def _on_reset_instance_action(self, event: ActionEvent) -> None:
         event.set_results(results)  # type: ignore[arg-type]
         self.model.unit.status = ops.ActiveStatus()
 
+    def _get_random_password(self) -> str:
+        """Get random password. Extracted from postgresql-k8s charm.
+
+        Returns:
+            random password.
+        """
+        choices = string.ascii_letters + string.digits
+        password = "".join([secrets.choice(choices) for i in range(16)])
+        return password
+
     def _on_register_user_action(self, event: ActionEvent) -> None:
         """Reset instance and report action result.
 
         Args:
             event: Event triggering the reset instance action.
         """
-        results = {
-            "register-user": False,
+        results = {"register-user": False, "user-password": ""}
+        container = self.unit.get_container(SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self.unit.status = ops.MaintenanceStatus("Waiting for pebble")
+            return
+        registration_shared_secret = self._synapse.get_configuration_field(
+            container=container, fieldname="registration_shared_secret"
+        )
+        if registration_shared_secret is None:
+            event.fail("registration_shared_secret was not found")
+            return
+        user_data = {
```

Comment:
This doesn't feel like it is in the right place, @gregory-schiano? Perhaps it should go into a separate user management module or the Synapse API can take care of it including the password generation?

Reply:
Update: I moved the password generation to User class.

Reply:
Update 1: all interactions with Synapse were moved to synapse module as suggested by @gregory-schiano thanks :)

Reply:
I'll mark this as resolved since the code has changed a lot, feel free to reopen it.